### PR TITLE
Prevent link highlight in markdown code blocks and spans

### LINKED ISSUE #139770: Link highlighting is applied inside codeblocks
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.63.2
- OS Version: Windows_NT 10.0.19043

Steps to Reproduce:

1. Open a .md file (or manually select md highlighting)
2. Enter:
````
```
[](a)
```
````

### PR DESCRIPTION
**Pull request by WaqasAliAbbasi**
This PR fixes #139770. There were four cases to look out for:
1. Indented code blocks
2. Fenced code blocks (surrounded ` ``` ` or `~~~` ) 
3. Inline code spans
   1. With backticks inside
   2. With newlines inside
   3. More than one backtick at beginning/end
   4. Cannot be preceded/followed with backticks

1 and 2 were solved by getting line numbers from https://github.com/microsoft/vscode/blob/2a0d371ca3a7c216185eb8f4bc3831420b3c9d1c/extensions/markdown-language-features/src/markdownEngine.ts#L94

For 3, the markdown engine doesn't provide character positions for inline elements. Hence I expanded the regex to capture inline code spans as well. A capture of inline code span takes precedence over link. If an inline code span is matched, we simply ignore it. Inline code span regex can be tested [here](https://regexr.com/6dd2h) with examples from [CommonMark Spec](https://spec.commonmark.org/0.30/#code-spans).

## Issues

1. ``[not a `link](/foo`)`` is still detected as a valid link because the entire line match takes precedence over codespan match.
2. Multiline Inline code span still shows up underlined in editor but for different reasons. (DocumentLinkProvider doesn't return it as link and we don't see Follow Link on hovering)
image

### CODE REVIEW

File: extensions/markdown-language-features/src/features/documentLinkProvider.ts [hunk 1]
```diff
@@ -102,18 +104,57 @@ export function stripAngleBrackets(link: string) {
...
+const codeSpanAndLinkPattern = /(?:(?<!`)(`+)(?!`)(?:.+?|.*?(?:(?:\r?\n).+?)*?)(?:\r?\n)?(?<!`)\1(?!`))|(?:(\[((!\[[^\]]*?\]\(\s*)([^\s\(\)]+?)\s*\)\]|(?:\\\]|[^\]])*\])\(\s*)(([^\s\(\)]|\([^\s\(\)]*?\))+)\s*(".*?")?\))/g;
 const referenceLinkPattern = /(\[((?:\\\]|[^\]])+)\]\[\s*?)([^\s\]]*?)\]/g;
 const definitionPattern = /^([\t ]*\[(?!\^)((?:\\\]|[^\]])+)\]:\s*)([^<]\S*|<[^>]+>)/gm;
 
+/**
+ * Asserts if number is inside at least one interval in the array of intervals.
+ *
+ * @param intervals An array of [a,b) sorted by `a` in ascending order.
+ */
+const isNumberInIntervals = (intervals: [number, number][], start: number, end: number, target: number): Boolean => {
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/documentLinkProvider.ts:116]: I know binary searches are theoretically better, but honestly I'd just use `some`:

```js
intervals.some(pair => target >= pair[0] && target < pair[1]);
```

Even if you have a document with 1000 code blocks, that will be plenty fast

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for looking into this! The change looks good overall. Just a few general comments

**Comment by WaqasAliAbbasi** [comment]
> Thanks for looking into this! The change looks good overall. Just a few general comments

Thanks for the useful feedback, what do you think of the changes?

**Comment by mjbvz** [comment]
Thanks for making these changes! LGTM. Just waiting for CI to run and then I will merge if everything looks good

**Comment by WaqasAliAbbasi** [comment]
> Thanks for making these changes! LGTM. Just waiting for CI to run and then I will merge if everything looks good

Seems like the CI is green, let's get it merged?

**Pull request merged**